stl/inc/algorithm

```diff
@@ -8077,6 +8094,85 @@ void partial_sort(_ExPo&&, _RanIt _First, _RanIt _Mid, _RanIt _Last) noexcept /*
     // parallelism suspected to be infeasible
     return _STD partial_sort(_First, _Mid, _Last);
 }
+
+#ifdef __cpp_lib_concepts
+namespace ranges {
+    // VARIABLE ranges::partial_sort
+    class _Partial_sort_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        // clang-format off
+        template <random_access_iterator _It, sentinel_for<_It> _Se, class _Pr = ranges::less, class _Pj = identity>
+            requires sortable<_It, _Pr, _Pj>
+        constexpr _It operator()(_It _First, _It _Mid, _Se _Last, _Pr _Pred = {}, _Pj _Proj = {}) const {
+            // clang-format on
+            _Adl_verify_range(_First, _Mid);
+            _Adl_verify_range(_Mid, _Last);
+
+            if constexpr (is_same_v<_It, _Se>) {
+                _Partial_sort_common(_Get_unwrapped(_STD move(_First)), _Get_unwrapped(_STD move(_Mid)),
+                    _Get_unwrapped(_Last), _Pass_fn(_Pred), _Pass_fn(_Proj));

```

Comment:
`move`s are mandatory when handling single-pass iterators since they could be move-only, and discretionary for `forward` (and better) iterators. The Ranges code moves iterators when it can (1) to avoid forgetting to move when we must, (2) in hopes that perf may sometimes be better for moves than copies, and (3) to give dataflow hints ("This thing is moved-from here indicating that the code is done with it.").

In this case, the iterators are random-access so we don't _need_ to move them, and in fact moving from `_Last` here would be problematic since we `return _Last` on the next line.

Reply:
Thank you for the super detailed explanation :) I have saved this comment to refer back to in future range PRs :)